ENH: issue overflow warning when using `abs` on `np.int8(-128)`

Micky774 opened the pull request:
## Related Issues/PRs
Resolves #21289

## Changes
Checks condition `a == NPY_MIN_@NAME@` to determine whether an overflow error has occurred for `np.int8` type. See #21289 and https://github.com/numpy/numpy/pull/21188#discussion_r866518325 for reference.

## Comments
This is my first PR and it's not-quite-complete yet but I wanted to open it so that I could gather some feedback and assistance. In particular, I'm not sure if this change is sufficient, nor how to write tests for this that adhere to NumPy standards. Any help would be appreciated!

Micky774 committed:
ENH Issue overflow warning when using `abs` on `np.int8(-128)`

Micky774 committed:
Update numpy/core/src/umath/scalarmath.c.src

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

Micky774 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -587,6 +587,9 @@ static NPY_INLINE int
 @name@_ctype_absolute(@type@ a, @type@ *out)
 {
     *out = (a < 0 ? -a : a);
+    if(a < 0 && a == -a){
```
How should I go about testing this, e.g. what test file should I add this under? I'm very unfamiliar w/ the NumPy codebase 😅

seberg replied:
@Micky774 yeah, not sure that using `*out` was a good idea in any case.

Many files have a corresponding test file, the same is true for scalarmath.  Since its a C file, those are in `numpy/core/tests`.   There are similar tests there (maybe even one that can be parametrized to include this one).

WarrenWeckesser commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -587,6 +587,9 @@ static NPY_INLINE int
 @name@_ctype_absolute(@type@ a, @type@ *out)
 {
     *out = (a < 0 ? -a : a);
+    if (a < 0 && a == -a) {
```
My understanding is that if `a` has a signed integer type and `a` equals the minimum integer, the result of `-a` is *undefined*.  I don't know the best reference for this, but an example where you can find this stated is in Table I on page A:4 of https://www.cs.utah.edu/~regehr/papers/tosem15.pdf.

This code *might* work fine with all the C compilers that we use, but in general, undefined behavior should be avoided.

Perhaps another template loop variable could be added that holds the type-dependent minimum integer name (something from `limits.h` or its numpy equivalent, if numpy has something like it), and a direct equality test could be used.

WarrenWeckesser replied:
The expresson `a == -a` is also used in `@name@_ctype_divide()`.

seberg replied:
Fair, maybe we should just use it?  I am sure we have two complement integers hard-coded all over the place and since we will have tests, we should notice if a system comes along where things actually break...

WarrenWeckesser replied:
I pointed it out thinking we should fix it there, too :grin:.  But I don't know if we want to expand the scope of this PR that much.

Micky774 replied:
I incorporated the suggestion to use templating for `NPY_MIN_@NAME@` and can create a follow-up PR to address the same unsafe check in `@name_ctype_divide()` if that would be of interest -- that way we can keep this PR small and easy to review.

Micky774 committed:
Added safer check for overflow

Micky774 committed:
Simplified check

Micky774 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -582,11 +582,15 @@ static NPY_INLINE int
 @name@_ctype_absolute(@type@ a, @type@ *out)
 {
     *out = (a < 0 ? -a : a);
+    if (a < 0 && a == NPY_MIN_@NAME@) {
```
Good catch!

Micky774 committed:
Merge branch 'main' into absolute_byte_overflow

Micky774 committed:
Fixes/reenables test

Micky774 committed:
Added fperror check to `@type@_absolute`

Micky774 committed:
Linting

Micky774 commented:
Couple things I found that I have addressed in this PR (trying to avoid too large a scope):

1. There was no `fperror` handling for `absolute` in `scalarmath.c` so I added it to the `.src` file.
2. `signed` vs `unsigned` overflow tests were set up incorrectly. In particular, `test_scalar_unsigned_integer_overflow` was actually titled `test_scalar_signed_integer_overflow` and overwrote it (not sure if intentional?)
3. The `test_scalar_signed_integer_overflow` test fails on integer floor division, so I set that as `xfail` to deal with in a follow-up PR.

ganesh-k13 commented:
Just mentioning https://github.com/numpy/numpy/pull/21507 as it handles a few cases.

seberg commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -584,11 +584,15 @@ static NPY_INLINE int
     *out = (a < 0 ? -a : a);
+    if (a == NPY_MIN_@NAME@) {
+        return NPY_FPE_OVERFLOW;
+    }
```
This will work in practice I guess.  In principle you should likely avoid calculating `-a` since it is undefined behavior in that branch.

Micky774 replied:
Went ahead and put this check first, in which case I directly assign `*out = a` for consistent behavior. Let me know if you have any concerns :)

seberg reviewed (commented):
Thanks, seems like it should be pretty much ready, the main thing is that I would prefer omitting the unnecessary floating point flag handling.

Micky774 committed:
Merge branch 'main' into absolute_byte_overflow

Micky774 committed:
Removed extra FPError handling

Micky774 committed:
Merge branch 'main' into absolute_byte_overflow

seberg commented on numpy/core/tests/test_scalarmath.py:
```diff
@@ -898,7 +898,8 @@ def test_scalar_integer_operation_overflow(dtype, operation):
         lambda min, neg_1: min * neg_1,
-        lambda min, neg_1: min // neg_1], ids=["abs", "*", "//"])
+        pytest.param(lambda min, neg_1: min // neg_1,
+            marks=pytest.mark.xfail)], ids=["abs", "*", "//"])
```
Probably need to comment it out or skip this, because it is just too broken...

Micky774 replied:
At the risk of slightly expanding the scope of this PR, I went ahead and just avoided the actual error-causing computation in the integer division function, setting it to return `NPY_MIN_@NAME@` as I think matches `gcc` behavior (in lieu of true defined behavior, this is the next-best thing imo). The test runs now, and the behavior matches expectations I think.

seberg replied:
OK, I would like to backport these division fixes at some point.  But I am starting to suspect we should make a PR with those from scratch… (i.e. a backport only PR)

Micky774 committed:
Fixed integer division FP_OVERFLOW

Micky774 committed:
Typo

Micky774 committed:
Caught warnings for negating unsigned integers

seberg commented:
@ganesh-k13 maybe you want to review/finalize this?  @Micky774, unfortunately, I don't quite believe you have double-checked that the tests check everything as advertised :).  Can you please make sure we have tests for this new abs path (and the negative one for that matter).  Tests that fail without the changes applied.

seberg commented on numpy/core/tests/test_scalarmath.py:
```diff
@@ -911,7 +911,7 @@ def test_scalar_signed_integer_overflow(dtype, operation):
 
 @pytest.mark.parametrize("dtype", np.typecodes["UnsignedInteger"])
 @pytest.mark.xfail  # TODO: the check is quite simply missing!
```
```suggestion
```
OK, nvm trying to let you to dig, just remove this :).  I suppose with this gone, all "new" paths should be hit.
I had just noticed the xfail was still here, so that is bad.  But of course the main (abs) change is tested.

seberg replied:
Sorry, this was really not the PR at all... But, I still think we can remove the xfail now?

mattip commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -167,7 +167,7 @@ static NPY_INLINE int
     else if (b == -1 && a == NPY_MIN_@NAME@) {
-        *out = a / b;
+        *out = NPY_MIN_@NAME@;
         return NPY_FPE_OVERFLOW;
```
This is suspicious: is this line really not hit during tests?

ganesh-k13 replied:
Yep https://github.com/numpy/numpy/pull/21507 has tests that covers this.

seberg replied:
I think both PRs have tests for this, so suspect that it is just in-lining/optimization leading to coverage errors (that might be the first time I have seen coverage failing, but...).

@ganesh-k13 if you prefer it to not be included here, maybe just push the change yourself?  (We have to mark that one additional test as `skip` then probably.)  Otherwise, I am happy to merge as-is.

Yes, it would be nice to follow-up to expand the tests so that they raise the error, and there is still lots of little stuff to clear out in `multiplication`... But at this point it feels like iterating here is getting tricky.

mattip commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -1564,8 +1569,23 @@ static PyObject *
+            return NULL;
+        }
+        Py_XDECREF(errobj);
+    }
```
It is suspicious that these new lines are not marked as being tested

seberg replied:
Hmmm, right.  The first one is OK (I don't think it can be reasonably covered, it indicates an internal/critical error).

To hit the second one, we should also check for:
```
with np.errstate(over="raise"):
    with pytest.raises(FloatingPointError):
         operation
```

ganesh-k13 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -167,7 +167,7 @@ static NPY_INLINE int
 #if @neg@
     else if (b == -1 && a == NPY_MIN_@NAME@) {
-        *out = a / b;
+        *out = NPY_MIN_@NAME@;
```
I think https://github.com/numpy/numpy/pull/21507 will cover it so we can remove this here perhaps? All though I'm happy to rebase if it goes in first.

ganesh-k13 reviewed (commented):
Sorry I missed the ping Sebastian, yeah the changes look ok but seems like Azure pipelines shows tests failed for windows builds that need to be checked.

Micky774 committed:
Merge branch 'main' into absolute_byte_overflow

Micky774 committed:
Enabled tests

seberg commented:
Gah, this is a mess... The linux code (mostly) uses `npy_mul_with_overflow_@name@` which is probably great and correct *when the builtin is available*.  When it is not available, it seems just plain for signed inputs.
The overflow checks themselves seem all pretty bogus to be honest?  Just xfail the test?  We should fix it, but not here...  Maybe same with the division test (to be able to pull out the change as Ganesh said).

Sorry that this is a a bit of never-ending story, because these integer overflow checks are so spotty, right now.

Micky774 commented:
> Gah, this is a mess... The linux code (mostly) uses `npy_mul_with_overflow_@name@` which is probably great and correct _when the builtin is available_. When it is not available, it seems just plain +wrong for signed inputs. The overflow checks themselves seem all pretty bogus to be honest? Just xfail the test? We should fix it, but not here... Maybe same with the division test (to be able to pull out the change as Ganesh said).
> 
> Sorry that this is a a bit of never-ending story, because these integer overflow checks are so spotty, right now.

I just spent the last couple hours eventually finding that check and scratching my head at it, so I appreciate the confirmation that I'm not going insane XD

Yeah it seems there's some systemic revisions necessary here. I suppose the least-invasive, most-likely-to-not-idle result would be to:

1. Maintain the current changes for `abs`
2. Maintain the enabling of the signed-overflow test (via the rename)
3. Maintain the enabling of the unsigned-overflow test (via removal of `xfail`)
4. `xfail` the multiplication test on the signed-overflow test to defer to a later PR
5. Tentatively be okay w/ some lack of code coverage, depending on #21507

Does that sound about right?

And thanks to all of you for your patience with this 😄

Micky774 committed:
Marked `xfail` for multiplication due to error on Windows

seberg committed:
MAINT: Revert change to division (and mark test as skip)

This is because similar changes are already included in a different
PR that fixes up more places where division may have problems.

seberg commented:
OK, just added that commit myself, so that this is again focused only on `abs` and adding the FPE raising logic (as it also affects negative partially).

I am going to merge it once the tests pass, thanks @Micky774 and Ganesh.  @Micky774 it would be cool to look into the open issues here:
* `-min_signed_int` was missing the check I think?
* It would be cool if those tests (well my failed attempt to write some ;)), gets fixed up to also run the code with `np.errstate(all="raise")` to cover the error path.
* I really would like to fix up the multiplication, that requires a bit of thought and its own PR though!  (not as important as fixing the division stuff, but also somewhat important NEP 50!).

seberg commented:
Only the `ppc64le` job isn't finished yet not going to wait for that.  Thanks.

seberg commented:
xref gh-21506 (for an eventual release note summarizing these).
